javascript/app-event-analysis/src/hooks/useCalculateEstimate.ts

```diff
@@ -0,0 +1,166 @@
+		isCalculatingEstimator,
+	)
+
+	return useCallback(
+		async (placebo = false, treatedUnitsList: string[] = treatedUnits) => {
+			// NOTE that a long running placebo can force the user to wait a long time
+			//  alternatively, a task queue mechanism can be used on the backend to schedule
+			//  long-running tasks on the background and let the frontend poll for the result as needed
+			//  See Celery as an example: https://testdriven.io/courses/fastapi-celery/intro/
+
+			if (!checkCanExecuteEstimator()) return
+
+			setUserMessage({
+				isVisible: true,
+				content: 'Calculating...',
+				type: MessageBarType.info,
+			})
+
+			setPlaceboSimulation(placebo)
+			setIsLoading(true)
+
+			// filter input data according to unit selection, and include the treated units
+			let filteredInputData = data.dataPoints
+				.filter(
+					dp =>
+						(checkedUnits && checkedUnits.has(dp.unit)) ||
+						treatedUnitsMap[dp.unit],
+				)
+				.map(({ unit, date, value, treated }) => {
+					// Make sure data point only includes value, date, unit, treated field
+					return { unit, date, value, treated }
+				})
+
+			if (placebo) {
+				filteredInputData = filteredInputData.filter(
+					i => i.unit === treatedUnitsList[0] || !treatedUnits.includes(i.unit),
+				)
+			}
+
+			// basically any changes to units/dates is considered placebo (i.e., what-if simulation)
+			// except when only excluding some of the units from the input data
+
+			// @TEMP use the following to switch the backend lib that is used
+			//       for calculating the estimate
+			const queryParams = new URLSearchParams(window.location.search)
+			const lib = queryParams.get('lib')
+
+			let calculationError: Error | null = null
+			const res = await API.post<SDIDOutputResponse>('/', {
+				input_data: filteredInputData,
+				estimator: estimator.toLowerCase(),
+				time_alignment: timeAlignment.toLowerCase(),
+				treatment_start_dates: treatmentStartDates,
+				compute_placebos: placebo,
+				treated_units: treatedUnitsList,
+				lib: lib === 'pysynthdid' ? lib : 'synthdid',
+			}).catch((error: Error) => {
+				calculationError = error
+			})
+
+			setIsLoading(false)
+
+			// result will always come back as an array
+			const outputsResponse: SDIDOutputResponse | null =
+				calculationError === null && res ? res.data : null
+			const outputs =
+				outputsResponse?.outputs.map(outputResult => outputResult.output) ?? []
+			if (outputs.every(output => isEmpty(output))) {
+				let errorMsg = 'Error calculating output for the provided input!'
+				if (calculationError !== null)
+					errorMsg += '\n' + (calculationError as Error).message
+				setUserMessage({
+					isVisible: true,
+					content: errorMsg,
+					type: MessageBarType.error,
+				})
+				setOutputRes(null)
+				setPlaceboOutputRes(new Map())
+				return
+			}
+			setUserMessage({
+				isVisible: false,
+				content: '',
+			})
+			if (placebo) {
+				setPlaceboOutputRes(prev => {
```

Comment:
Is a `map` the right object for this? I think maps are intended to be mutable by default. If we're creating fresh instances on mutation, maybe a `Record<key,value>` makes more sense